direct/src/gui/DirectFrame.py

```diff
@@ -50,166 +52,113 @@ def __init__(self, parent = None, **kw):
-                if text == None:
-                    return
-                else:
-                    from .OnscreenText import OnscreenText
-                    self.createcomponent(
-                        component, (), 'text',
-                        OnscreenText,
-                        (), parent = self.stateNodePath[i],
-                        text = text, scale = 1, mayChange = self['textMayChange'],
-                        sort = DGG.TEXT_SORT_INDEX,
-                        )
+        self.new_set("text", text)
 
     def clearGeom(self):
         self['geom'] = None
         self.setGeom()
 
-    def setGeom(self, geom=None):
-        if geom is not None:
-            self['geom'] = geom
-
-        # Determine argument type
-        geom = self['geom']
-
-        if geom == None:
-            # Passed in None
-            geomList = (None,) * self['numStates']
-        elif isinstance(geom, NodePath) or \
-             isinstance(geom, stringType):
-            # Passed in a single node path, make a tuple out of it
-            geomList = (geom,) * self['numStates']
+    def long_type_check(self, name, object_):
+        if isinstance(object_, stringType):
+            return True
+        if name == "image":
+            if ((isinstance(object_, NodePath) or
+                isinstance(object_, Texture)) 
+                or
+                ((len(object_) == 2) and# Its a model/node pair of strings
+                isinstance(object_[0], stringType) and
+                isinstance(object_[1], stringType))):
+                return True
+        elif name == "geom":
+            if (isinstance(object_, NodePath)):
+                return True
+        return False
+
+    def new_set(self, name, object_=None):
+        """my new code golf par is 1"""
+        assert name in ["geom", "image", "text"]
+        if object_ is not None:
+            self[name] = object_
+
+        object_ = self[name]
+
+        if object_ == None:
+            object_list = (None,) * self['numStates']
         else:
-            # Otherwise, hope that the user has passed in a tuple/list
-            geomList = geom
-
-        # Create/destroy components
-        for i in range(self['numStates']):
-            component = 'geom' + repr(i)
-            # If fewer items specified than numStates,
-            # just repeat last item
-            try:
-                geom = geomList[i]
-            except IndexError:
-                geom = geomList[-1]
-
-            if self.hascomponent(component):
-                if geom == None:
-                    # Destroy component
-                    self.destroycomponent(component)
+            if self.long_type_check(name, object_):
+                object_list = (object_,) * self['numStates']
+            else:
+                object_list = object_
+
+        name_based_classes = {
+            "geom": OnscreenGeom,
+            "image": OnscreenImage,
+            "text": OnscreenText, }
```

Comment:
I don't like these lookups.  It's a lot more elegant to simply pass this as argument to the `new_set` helper.  Same for `name_based_constants`, below.